geomstats/geometry/sasaki_metric.py

```diff
@@ -0,0 +1,238 @@
+
+        Parameters
+        ----------
+        tangent_vec : array-like, shape=[..., 2, M.dim]
+            Tangent vector in TTM at the base point in TM.
+        base_point : array-like, shape=[..., 2, M.dim]
+            Point in the tangent bundle TM of manifold M.
+        n_steps : int
+            Number of discrete time steps.
+            Optional, default: self.n_s.
+
+        Returns
+        -------
+        exp : array-like, shape=[..., 2, M.dim]
+            Point on the tangent bundle TM.
+        """
+        # unflatten
+        bs_pts = gs.reshape(base_point, (-1, 2) + self.metric.shape)
+        tngs = gs.reshape(tangent_vec, bs_pts.shape)
+
+        metric = self.metric
+        par_trans = metric.parallel_transport
+        n_s = self.n_s if n_steps is None else n_steps
+        eps = 1 / n_s
+
+        v0, w0 = tngs[:, 0], tngs[:, 1]
+        p0, u0 = bs_pts[:, 0], bs_pts[:, 1]
+        p, u = p0, u0
+        for _ in range(n_s):
+            p = metric.exp(eps * v0, p0)
+            u = par_trans(u0 + eps * w0, p0, None, p)
+            v = par_trans(v0 - eps * (metric.curvature(u0, w0, v0, p0)), p0, None, p)
+            w = par_trans(w0, p0, None, p)
+            p0, u0 = p, u
+            v0, w0 = v, w
+
+        return gs.reshape(gs.array(list(zip(p, u))), base_point.shape)
+
+    def log(self, point, base_point, n_steps=None, **kwargs):
+        """Compute the Riemannian logarithm of a point.
+
+        Logarithmic map at base_point of point computed by
+        iteratively relaxing a discretized geodesic between base_point and point.
+
+        Parameters
+        ----------
+        point : array-like, shape=[..., 2, M.dim]
+            Point in the tangent bundle TM of manifold M.
+        base_point : array-like, shape=[..., 2, M.dim]
+            Point in the tangent bundle TM of manifold M.
+        n_steps : int
+            Number of discrete time steps.
+            Optional, default: self.n_s.
+
+        Returns
+        -------
+        log : array-like, shape=[..., 2, M.dim]
+            Tangent vector at the base point equal to the Riemannian logarithm
+            of point at the base point.
+        """
+        # unflatten
+        pts = gs.reshape(point, (-1, 2) + self.metric.shape)
+        bs_pts = gs.reshape(base_point, (-1, 2) + self.metric.shape)
+
+        metric = self.metric
+        par_trans = metric.parallel_transport
+        n_s = self.n_s if n_steps is None else n_steps
+
+        @delayed
+        def do_log(pt, bs_pt):
+            """Calculate the logarithm."""
+            pu = self.geodesic_discrete(bs_pt, pt, n_s)
+            p1, u1 = pu[1][0], pu[1][1]
+            p0, u0 = bs_pt[0], bs_pt[1]
+            w = (par_trans(u1, p1, None, p0) - u0)
+            v = metric.log(point=p1, base_point=p0)
+            return n_s * gs.array([v, w])
+
+        n_jobs = min(os.cpu_count(), len(pts))
+        with Parallel(n_jobs=n_jobs, verbose=0) as parallel:
```

Comment:
We think the is the right place for parallelization as the loop here is potentially way larger (hundreds/thousands) as they depend on the size of your dataset, if you e.g. consider statistical operations. Whereas in the `geodesic_discrete` we have a default n_steps of 3.

Reply:
Ah, ok, now I got your point. You basically suggest to wrapper the whole `geodesic_discrete`  s.t. it becomes vectorized through an interal parallel loop.

Reply:
My main idea here is exactly what you refer: having a parallelized internal loop. Why?
* the parallelism is placed exactly where it belongs, i.e. in the optimization problem for `geodesic_discrete`.
* `geodesic_discrete` starts accepting multiple `end_point` and `initial_point`. So if for some reason we want to compute several discrete geodesics, then we know the parallelization is taken care of for us.
* we can vectorize operations in `log` after getting the discrete geodesics (e.g. parallel transport and metric.log), which may be much faster if we have tons of points.

Why do you think it may be a problem for later polishing? btw, we are thinking about a change in code architecture (#1152) that will play a role here.

Reply:
> Why do you think it may be a problem for later polishing?

I think my sentence was a little misleading and I don't think this is an issue for later polishing. Instead I meant that it is something that could be done later when performing code polishing. However, I will put in the requested changes.